src/ast.ts

```diff
@@ -107,54 +183,234 @@ export function collectFields(
 }
 
 /**
- * Determines if a field should be included based on the @include and @skip
- * directives, where @skip has higher precedence than @include.
+ * Augment __internalShouldInclude code for all sub-fields in the
+ * tree with @param rootfieldNode as the root.
+ *
+ * This is required to handle cases where there are multiple paths to
+ * the same node. And each of those paths contain different skip/include
+ * values.
+ *
+ * For example,
+ *
+ * ```
+ * {
+ *   foo @skip(if: $c1) {
+ *     bar @skip(if: $c2)
+ *   }
+ *   ... {
+ *     foo @skip(if: $c3) {
+ *       bar
+ *     }
+ *   }
+ * }
+ * ```
+ *
+ * We decide shouldInclude at runtime per fieldNode. When we handle the
+ * field `foo`, the logic is straight forward - it requires one of $c1 or $c3
+ * to be false.
+ *
+ * But, when we handle the field `bar`, and we are in the context of the fieldNode,
+ * not enough information is available. This is because, if we only included $c2
+ * to decide if bar is included, consider the case -
+ *
+ * $c1: true, $c2: true, $c3: false
+ *
+ * If we considered only $c2, we would have skipped bar. But the correct implementation
+ * is to include bar, because foo($c3) { bar } is not skipped. The entire sub-tree's
+ * logic is required to handle bar.
+ *
+ * So, to handle this case, we augment the tree at each point to consider the
+ * skip/include logic from the parent as well.
+ *
+ * @param compilationContext {CompilationContext} Required for getFragment by
+ * name to handle fragment spread operation.
+ *
+ * @param rootFieldNode {JitFieldNode} The root field to traverse from for
+ * adding __internalShouldInclude to all sub field nodes.
  */
-function shouldIncludeNode(
-  exeContext: ExecutionContext,
-  node: FragmentSpreadNode | FieldNode | InlineFragmentNode
-): boolean {
-  const skip = getDirectiveValues(
-    GraphQLSkipDirective,
-    node,
-    exeContext.variableValues
-  );
-  if (skip && skip.if === true) {
-    return false;
+function augmentFieldNodeTree(
+  compilationContext: CompilationContext,
+  rootFieldNode: JitFieldNode
+) {
+  for (const selection of rootFieldNode.selectionSet?.selections ?? []) {
+    handle(selection);
   }
 
-  const include = getDirectiveValues(
-    GraphQLIncludeDirective,
-    node,
-    exeContext.variableValues
+  /**
+   * Recursively traverse through sub-selection and combine `shouldInclude`s
+   * from parent and current ones.
+   */
+  function handle(selection: SelectionNode) {
+    switch (selection.kind) {
+      case Kind.FIELD: {
+        const jitFieldNode: JitFieldNode = selection;
+        jitFieldNode.__internalShouldInclude = joinShouldIncludeCompilations(
+          rootFieldNode.__internalShouldInclude ?? "",
```

Comment:
this should be parentFieldNode

Reply:
The main question is why a test is not broken. :wink: